File: currency/src/main/java/ucb/arqsoft/currency/api/CurrencyApi.java
package ucb.arqsoft.currency.api;

import java.math.BigDecimal;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import ucb.arqsoft.currency.bl.CurrencyBl;
import ucb.arqsoft.currency.dto.ExchangeDto;
import ucb.arqsoft.currency.dto.ResponseDto;

@RestController
@RequestMapping("/api/v1/currency")
public class CurrencyApi {

    private static Logger logger = LoggerFactory.getLogger(CurrencyApi.class);
    private CurrencyBl currencyBl;

    public CurrencyApi(CurrencyBl currencyBl) {
        this.currencyBl = currencyBl;
    }

    @GetMapping
    public ResponseEntity<ResponseDto<ExchangeDto>> getExchange(
        @RequestParam String from,
        @RequestParam String to,
        @RequestParam BigDecimal amount
    ) {
        logger.info("GET: Exchange " +  amount + " " + from + " to " + to);
        logger.info("Starting business logic");
        ExchangeDto exchangeDto = currencyBl.exchange(from, to, amount);
        ResponseDto<ExchangeDto> responseDto = new ResponseDto<ExchangeDto>(exchangeDto, true, null);
        logger.info("Sending response");
        return ResponseEntity.ok(responseDto);
    }

}

File: currencykt/src/main/kotlin/ucb/arqsoft/currencykt/api/CurrencyApi.kt
package ucb.arqsoft.currencykt.api

import okhttp3.internal.connection.Exchange
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import ucb.arqsoft.currencykt.bl.CurrencyBl
import ucb.arqsoft.currencykt.dto.ExchangeDto
import ucb.arqsoft.currencykt.dto.PaginatedDto
import ucb.arqsoft.currencykt.dto.ResponseDto
import java.math.BigDecimal

@RestController
@RequestMapping("/api/v1/currency")
class CurrencyApi (private val currencyBl: CurrencyBl) {

    private val logger: Logger = LoggerFactory.getLogger(CurrencyApi::class.java)

    @GetMapping
    fun getCurrency(
        @RequestParam from: String,
        @RequestParam to: String,
        @RequestParam amount: BigDecimal
    ): ResponseEntity<ResponseDto<ExchangeDto>> {
        logger.info("GET: Exchange $amount from $from to $to");
        logger.info("Starting business logic");
        val exchangeDto = currencyBl.exchangeRate(amount, from, to);

        return ResponseEntity.ok(
            ResponseDto(
                data = exchangeDto,
                message = "Success",
                successful = true
            )
        );
    }

    @GetMapping("/all")
    fun getAllExchanges(
        @RequestParam limit: Int,
        @RequestParam offset: Int,
        @RequestParam requestParams: Map<String, String>
    ): ResponseEntity<ResponseDto<PaginatedDto<ExchangeDto>>> {
        logger.info("GET: $limit exchanges starting from $offset");
        val paginatedExchangeDto: PaginatedDto<ExchangeDto> =
            currencyBl.getExchangeList(limit, offset, requestParams);
        logger.info("Sending response");
        return ResponseEntity.ok(
            ResponseDto(
                data = paginatedExchangeDto,
                message = "Success",
                successful = true
            )
        )
    }

}